packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/base/AtomicResult.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.amplify.amplify_auth_cognito.base
+
+import io.flutter.plugin.common.MethodChannel
+import kotlinx.coroutines.*
+
+/**
+ * Thread-safe [MethodChannel.Result] wrapper which prevents multiple replies and automatically posts
+ * results to the main thread.
+ */
+class AtomicResult(private val result: MethodChannel.Result) : MethodChannel.Result {
+    private companion object {
+        /**
+         * Scope for performing result handling.
+         * Method channel results must be sent on the main (UI) thread.
+         */
+        val scope = MainScope()
+    }
+
+    /**
+     * Whether a response has been sent.
+     */
+    private var isSent = false
+
+    override fun success(value: Any?) {
+        scope.launch {
+            if (isSent) return@launch
```

Comment:
This just silently ignores everything other than the first response, correct? I don't know that we want to silently ignore them. If we are getting multiple responses, in most cases it there is probably an issue with our code or an issue in the native library. I think we should at least log something.

Reply:
Right now we don't have a good solution for logging, although I totally agree it would be good to monitor. The most important thing, though, is that it doesn't submit multiple replies which will crash the app. It's pretty rare that I've seen it and usually the result of poorly written multithreaded code, so I don't expect it from the Amplify SDK or our own code. It's more a piece-of-mind for now. When I've finished the logging SDK (next sprint), I will make sure to capture that information here.

Reply:
We actually had to do a few fixes on the storage plugin which did something very similar, just not as elegantly as this.

Reply:
Good to know - yeah multi-threading is hard. Having these as a fallback is definitely peace of mind for me because I've written plenty of deadlocks and race conditions before.

Reply:
> Right now we don't have a good solution for logging

Can we at least log this locally for now?

Reply:
Added!